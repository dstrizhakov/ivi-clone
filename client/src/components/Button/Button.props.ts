import { ButtonHTMLAttributes, DetailedHTMLProps, ReactNode } from 'react';

export interface ButtonProps
  extends DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
  children: ReactNode | ReactNode[];
  appearance?: 'rectangle' | 'square' | 'circle' | 'red' | 'transparent';
  size?: 'S' | 'M' | 'L';
}
