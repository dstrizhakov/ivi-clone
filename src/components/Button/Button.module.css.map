{"version":3,"sourceRoot":"","sources":["Button.module.scss"],"names":[],"mappings":"AAAA;EACE;EACA;EACA;EACA;EACA;EACA;EACA;EACA;EACA;EACA;EACA;EACA;EACA;;AACA;EACE;;;AAIJ;EACE;;;AAGF;EACE;EACA;EACA;;;AAGF;EACE;EACA;EACA;;;AAGF;EACE;EACA;EACA;;AACA;EACE;EACA;;;AAIJ;EACE;EACA;EACA;;;AAGF;EACE;EACA;;;AAGF;EACE;EACA","file":"Button.module.css"}