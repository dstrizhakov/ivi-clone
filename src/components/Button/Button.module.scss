@import '../../styles/variables.scss';

.button {
  z-index: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 5px;
  cursor: pointer;
  transition: $time-main;
  height: 40px;
  padding: 0 12px;
  background-color: #1f1b2e;
  border-color: #1f1b2e;
  color: white;
  font-size: 15px;
  font-weight: 600;
  &:disabled {
    opacity: 0.4;
    cursor: default;
    &:hover {
      background-color: #1f1b2e;
    }
  }
  &:hover {
    background-color: #28233b;
  }
}

.rectangle {
  border-radius: 8px;
}

.square {
  width: 40px;
  height: 40px;
  border-radius: 8px;
}

.circle {
  width: 40px;
  height: 40px;
  border-radius: 20px;
}

.red {
  background-color: #ea003d;
  border-color: #ea003d;
  border-radius: 8px;
  &:disabled {
    opacity: 0.3;
    cursor: default;
    &:hover {
      background-color: #ea003d;
    }
  }
  &:hover {
    background-color: #ff0f4d;
    border-color: #ff0f4d;
  }
}

.S {
  height: 32px;
  padding: 7px 11px;
  font-size: 13px;
}

.M {
  height: 40px;
  padding: 9px 15px;
}

.L {
  height: 48px;
  padding: 5px 15px;
}

.transparent {
  padding: 0;
  background-color: transparent;
  border: 0;
  border-radius: 8px;
  &:hover {
    background-color: transparent;
    border: 0;
  }
}

.button {
  font-size: 0.8rem;
}
