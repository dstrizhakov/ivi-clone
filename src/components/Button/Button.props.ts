import { ButtonHTMLAttributes, DetailedHTMLProps, ReactNode } from 'react';

export enum BtnA {
  rectangle = 'rectangle',
  square = 'square',
  circle = 'circle',
  red = 'red',
  transparent = 'transparent',
}

export enum BtnS {
  S = 'S',
  M = 'M',
  L = 'L',
}

export interface ButtonProps
  extends DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
  children: ReactNode | ReactNode[];
  appearance?: BtnA; //'rectangle' | 'square' | 'circle' | 'red' | 'transparent';
  size?: BtnS; //'S' | 'M' | 'L'
}
