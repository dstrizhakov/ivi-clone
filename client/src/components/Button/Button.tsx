import { ButtonProps } from './Button.props';
import styles from './Button.module.scss';
import cn from 'classnames';

export const Button = ({
  children,
  appearance = 'rectangle',
  size = 'M',
  className,
  ...props
}: ButtonProps): JSX.Element => {
  return (
    <button
      className={cn(
        styles.button,
        className,
        {
          [styles.rectangle]: appearance == 'rectangle',
          [styles.square]: appearance == 'square',
          [styles.circle]: appearance == 'circle',
          [styles.red]: appearance == 'red',
          [styles.transparent]: appearance == 'transparent',
        },
        {
          [styles.S]: size == 'S',
          [styles.M]: size == 'M',
          [styles.L]: size == 'L',
        }
      )}
      {...props}
    >
      {children}
    </button>
  );
};
