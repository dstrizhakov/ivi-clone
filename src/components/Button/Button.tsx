import { BtnA, BtnS, ButtonProps } from './Button.props';
import styles from './Button.module.scss';
import cn from 'classnames';
import { FC } from 'react';

export const Button: FC<ButtonProps> = ({
  children,
  appearance = BtnA.rectangle,
  size = BtnS.M,
  className,
  ...props
}): JSX.Element => {
  return (
    <button
      className={cn(
        styles.button,
        className,
        {
          [styles.rectangle]: appearance == BtnA.rectangle,
          [styles.square]: appearance == BtnA.square,
          [styles.circle]: appearance == BtnA.circle,
          [styles.red]: appearance == BtnA.red,
          [styles.transparent]: appearance == BtnA.transparent,
        },
        {
          [styles.S]: size == BtnS.S,
          [styles.M]: size == BtnS.M,
          [styles.L]: size == BtnS.L,
        }
      )}
      {...props}
    >
      {children}
    </button>
  );
};
