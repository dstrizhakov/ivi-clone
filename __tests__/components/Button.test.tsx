import { Button } from '../../src/components/Button/Button';
import { render, screen } from '@testing-library/react';
import { BtnA } from '../../src/components/Button/Button.props';
import React from 'react';

describe('Button', () => {
  it('should be circle', () => {
    render(<Button appearance={BtnA.circle}>btn</Button>);
    const button = screen.getByRole('button', { name: /btn/i });
    expect(button).toHaveClass('circle');
    expect(button).toMatchSnapshot();
  });
  it('should be red', () => {
    render(<Button appearance={BtnA.red}>btn</Button>);
    const button = screen.getByRole('button', { name: /btn/i });
    expect(button).toHaveClass('red');
    expect(button).toMatchSnapshot();
  });
  it('should be square', () => {
    render(<Button appearance={BtnA.square}>btn</Button>);
    const button = screen.getByRole('button', { name: /btn/i });
    expect(button).toHaveClass('square');
    expect(button).toMatchSnapshot();
  });
  it('should be rectangle', () => {
    render(<Button appearance={BtnA.rectangle}>btn</Button>);
    const button = screen.getByRole('button', { name: /btn/i });
    expect(button).toHaveClass('rectangle');
    expect(button).toMatchSnapshot();
  });
  it('should be rectangle default', () => {
    render(<Button>btn</Button>);
    const button = screen.getByRole('button', { name: /btn/i });
    expect(button).toHaveClass('rectangle');
    expect(button).toMatchSnapshot();
  });
});
