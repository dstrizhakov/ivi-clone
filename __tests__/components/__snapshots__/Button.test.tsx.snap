// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should be circle 1`] = `
<button
  class="button circle M"
>
  btn
</button>
`;

exports[`Button should be rectangle 1`] = `
<button
  class="button rectangle M"
>
  btn
</button>
`;

exports[`Button should be rectangle default 1`] = `
<button
  class="button rectangle M"
>
  btn
</button>
`;

exports[`Button should be red 1`] = `
<button
  class="button red M"
>
  btn
</button>
`;

exports[`Button should be square 1`] = `
<button
  class="button square M"
>
  btn
</button>
`;
